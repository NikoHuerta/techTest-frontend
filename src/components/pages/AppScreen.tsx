import { AppLayout } from '../layout';
import { useAppSelector } from '../../hooks';
import { HomeScreen, ViewTaskScreen, NewTaskScreen, EditTaskScreen } from './';


export const AppScreen = () => {

  const page = useAppSelector(state => state.navigation.status);

  return (
    <>
        <AppLayout>
            { 
                page === 'allTasks' ? <ViewTaskScreen />
                : page === 'newTask' ? <NewTaskScreen />
                : page === 'editTask' ? <EditTaskScreen />
                : <HomeScreen />
            }
        </AppLayout>
    </>
  )
}
