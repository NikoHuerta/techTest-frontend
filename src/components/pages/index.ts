export * from './HomeScreen';
export * from './AppScreen';
export * from './NewTaskScreen';
export * from './ViewTaskScreen';
export * from './EditTaskScreen';