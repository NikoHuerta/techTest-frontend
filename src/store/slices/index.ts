export { default as taskReducer } from './taskSlice';
export { default as navigationReducer } from './navigationSlice';
export * from './navigationSlice';
export * from './taskSlice';