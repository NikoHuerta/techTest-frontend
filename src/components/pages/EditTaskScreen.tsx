import { ChangeEvent, FormEvent, useState } from 'react'; 

import { Box, Button, FormControlLabel, Grid, TextField, Checkbox } from '@mui/material';
import { useSnackbar } from 'notistack';
import moment from 'moment';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';
import { DateTimePicker } from '@mui/x-date-pickers/DateTimePicker';

import { useAppDispatch, useAppSelector } from '../../hooks';
import { setNavigationStatus, updateTask } from '../../store/slices';
import { ITask } from '../../interfaces';

export const EditTaskScreen = () => {

  const { enqueueSnackbar } = useSnackbar();
  const dispatch = useAppDispatch();
  const { objectId, descripcion, fechaCreacion, vigente } = useAppSelector(state => state.task.activeTask!);


  const [valueDescription, setValueDescription] = useState<string>(descripcion);
  const [touched, setTouched] = useState(false);
  const [valueDate, setValueDate] = useState<Date | null>( new Date(fechaCreacion) );
  const [valueVigente, setVigente] = useState<boolean | null>(vigente);


  const handleDateChange = (newValue: Date | null) => {
    setValueDate(newValue);
  };

  const handleVigenteChange = (event: ChangeEvent<HTMLInputElement>) => {
    setVigente(event.target.checked);
  };

  const handleDescriptionChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValueDescription(event.target.value);
  };

  const handleUpdate = async (e: FormEvent<HTMLFormElement>) =>{
    e.preventDefault();
    if(valueDescription ==='') {
      return enqueueSnackbar('La descripcion no puede ser vacia', { 
        variant: 'error',
        autoHideDuration: 2000,
        anchorOrigin: {
          vertical: 'top',
          horizontal: 'right',
        } 
      });
    }
    
    const updatedTask: ITask = {
      descripcion : (valueDescription) ? valueDescription : '',
      fechaCreacion: moment(valueDate).format('YYYY-MM-DD HH:mm:ss'),
      vigente : (valueVigente) ? valueVigente : false,
      objectId,
    }

    try {
      const resp = await dispatch(updateTask(updatedTask));

      if(resp.meta.requestStatus === 'fulfilled') {
        enqueueSnackbar('Tarea editada correctamente', { 
          variant: 'success',
          autoHideDuration: 2000,
          anchorOrigin: {
            vertical: 'top',
            horizontal: 'right',
          } 
        });
      } else {
        enqueueSnackbar('Error en la comunicacion con backend', { 
          variant: 'error',
          autoHideDuration: 2000,
          anchorOrigin: {
            vertical: 'top',
            horizontal: 'right',
          } 
        });
      }

    } catch (e:any) {
      console.log('error', e);
    }
  }


  return (
    <>
      <h1>Edit Task Screen</h1>


      <form onSubmit={ handleUpdate } noValidate >
          
          <Grid container spacing={2} sx={{ mt:5 }}>
            <Grid item xs={6} sm={4}>
              <TextField
                label='Descripcion'
                name='descripcion'
                variant='filled'
                fullWidth
                required

                helperText={ valueDescription!.length === 0 && touched && 'Ingrese un valor'}
                error={ valueDescription!.length === 0 && touched }
                value={ valueDescription }
                onChange={ handleDescriptionChange }
                onBlur={ () => setTouched(true) }


              />
            </Grid>
            <Grid item xs={6} sm={4}>
              <LocalizationProvider dateAdapter={ AdapterMoment }>
                <DateTimePicker
                  label="Fecha de Creacion"
                  value={ valueDate }
                  onChange={ handleDateChange }
                  renderInput={( params ) => <TextField fullWidth variant='filled' name='fechaCreacion' {...params} />}
                />
              </LocalizationProvider>
              
            </Grid>  
            <Grid item xs={6} sm={4}>
              <FormControlLabel
                control={ 
                  <Checkbox checked={ valueVigente! } onChange={ handleVigenteChange } name='vigente' />
                }
                label='Vigente'
              />
            </Grid>    
          </Grid>

          <Box sx={{ mt: 5 }} display='flex' justifyContent='center'>
            
            <Button 
                type='button'
                color='primary' 
                className='circular-btn' 
                size='large'
                onClick={ () => { dispatch(setNavigationStatus({ status: 'allTasks' })) } }
              >
                Volver
            </Button>

            <Button 
              type='submit'
              color='primary' 
              className='circular-btn' 
              size='large'
            >
              Actualizar Tarea
            </Button>
          </Box>
      
      </form>

    </>
  )
}
