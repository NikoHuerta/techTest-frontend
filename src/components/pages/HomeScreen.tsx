import { Typography } from "@mui/material";
import { useEffect } from "react";
import { useAppDispatch } from "../../hooks";
import { loadTasks } from "../../store/slices";

export const HomeScreen = () => {

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(loadTasks());
  }, [dispatch]);
  
  
  return (
    <>
      <h1>Home Screen</h1>

      <Typography variant='subtitle1' gutterBottom sx={{ my: 5 }}>
        App encargada de manejar tareas. 
      </Typography>

      <Typography variant='subtitle2' gutterBottom sx={{ mb: 2 }}>
        Para ver las tareas, haga click en el boton "Ver Tareas".
      </Typography>

      <Typography variant='subtitle2' gutterBottom>
        Para agregar una nueva tarea, haga click en el boton "Nueva Tarea".
      </Typography>
      <Typography variant='subtitle2' gutterBottom>
        En la visualizacion usted podra Editar y Borrar una tarea respectiva.
      </Typography>
    </>
  )
}