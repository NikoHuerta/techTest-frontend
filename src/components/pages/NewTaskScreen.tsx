import { ChangeEvent, FormEvent, useState } from 'react'; 

import { Box, Button, FormControlLabel, Grid, TextField, Checkbox } from '@mui/material';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';
import { DateTimePicker } from '@mui/x-date-pickers/DateTimePicker';
import { useSnackbar } from 'notistack';
import moment from 'moment';

import { useAppDispatch } from '../../hooks';
import { addTask } from '../../store/slices';
import { ITask } from '../../interfaces/task';



export const NewTaskScreen = () => {

  const { enqueueSnackbar } = useSnackbar();
  const dispatch = useAppDispatch();

  const [valueDescription, setValueDescription] = useState<string | null>('');
  const [touched, setTouched] = useState(false);
  const [valueDate, setValueDate] = useState<Date | null>( new Date(moment.now()) );
  const [valueVigente, setVigente] = useState<boolean | null>(true);

  const handleDateChange = (newValue: Date | null) => {
    setValueDate(newValue);
  };

  const handleVigenteChange = (event: ChangeEvent<HTMLInputElement>) => {
    setVigente(event.target.checked);
  };

  const handleDescriptionChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValueDescription(event.target.value);
  };

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) =>{
    e.preventDefault();
    if(valueDescription ==='') {
      return enqueueSnackbar('Debe ingresar una descripcion', { 
        variant: 'error',
        autoHideDuration: 2000,
        anchorOrigin: {
          vertical: 'top',
          horizontal: 'right',
        } 
      });
    } 

    const newTask: ITask = {
      descripcion : (valueDescription) ? valueDescription : '',
      fechaCreacion: moment(valueDate).format('YYYY-MM-DD HH:mm:ss'),
      vigente : (valueVigente) ? valueVigente : false,
    }

    try {
      const resp = await dispatch(addTask(newTask));

      if(resp.meta.requestStatus === 'fulfilled') {
        enqueueSnackbar('Tarea agregada correctamente', { 
          variant: 'success',
          autoHideDuration: 2000,
          anchorOrigin: {
            vertical: 'top',
            horizontal: 'right',
          } 
        });
      } else {
        enqueueSnackbar('Error en la comunicacion con backend', { 
          variant: 'error',
          autoHideDuration: 2000,
          anchorOrigin: {
            vertical: 'top',
            horizontal: 'right',
          } 
        });
      }

    } catch (e:any) {
      console.log('error', e);
    }
      

  }


  return (
    <>
        <h1>New Task Screen</h1>
        
        <form onSubmit={ handleSubmit } noValidate >
          <Grid container spacing={2} sx={{ mt:5 }}>
            <Grid item xs={6} sm={4}>
              <TextField
                label='Descripcion'
                name='descripcion'
                variant='filled'
                fullWidth
                required

                helperText={ valueDescription!.length === 0 && touched && 'Ingrese un valor'}
                error={ valueDescription!.length === 0 && touched }
                value={ valueDescription }
                onChange={ handleDescriptionChange }
                onBlur={ () => setTouched(true) }


              />
            </Grid>
            <Grid item xs={6} sm={4}>
              <LocalizationProvider dateAdapter={ AdapterMoment }>
                <DateTimePicker
                  label="Fecha de Creacion"
                  value={ valueDate }
                  onChange={ handleDateChange }
                  renderInput={( params ) => <TextField fullWidth variant='filled' name='fechaCreacion' {...params} />}
                />
              </LocalizationProvider>
              
            </Grid>  
            <Grid item xs={6} sm={4}>
              <FormControlLabel
                control={ 
                  <Checkbox checked={ valueVigente! } onChange={ handleVigenteChange } name='vigente' />
                }
                label='Vigente'
              />
            </Grid>    
          </Grid>

          <Box sx={{ mt: 5 }} display='flex' justifyContent='center'>
            <Button 
              type='submit'
              color='primary' 
              className='circular-btn' 
              size='large'
            >
              Crear Tarea
            </Button>
          </Box>


        </form>
    </>
  )
}
