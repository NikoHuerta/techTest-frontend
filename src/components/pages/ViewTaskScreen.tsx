import { Button, Chip, Grid } from '@mui/material';
import { DataGrid, GridColDef, GridValueGetterParams } from '@mui/x-data-grid';
import ModeEditOutlineIcon from '@mui/icons-material/ModeEditOutline';
import DeleteOutlineIcon from '@mui/icons-material/DeleteOutline';
import moment from 'moment';

import { useAppDispatch, useAppSelector } from '../../hooks';
import { setActiveTask, setNavigationStatus } from '../../store/slices';


export const ViewTaskScreen = () => {

  const dataRows = useAppSelector(state => state.task.tasks);
  const dispatch = useAppDispatch();

  const rows = dataRows.map( (task, index) => {
    return {
      id: index,
      num: index + 1,
      descripcion: task.descripcion,
      fechaCreacion: moment(task.fechaCreacion).format('MMMM Do YYYY, h:mm'),
      vigente: task.vigente,
      objectId: task.objectId
    }
  });

  const columns: GridColDef[] = [
      { field: 'num', headerName: 'Num', width: 100, sortable: true },
      { field: 'descripcion', headerName: 'Descripcion', width: 200, sortable: true },
      { field: 'fechaCreacion', headerName: 'Fecha de Creacion', width: 300, sortable: true },
      { 
        field: 'vigente', 
        headerName: 'Vigente',
        renderCell: ({ row }: GridValueGetterParams) => {
          return row.vigente 
            ? (<Chip variant='outlined' label='Si' color='success' />) 
            : (<Chip variant='outlined' label='No' color='error' />)
        },
        width: 100, 
        sortable: true 
      },
      {
        field: 'actions',
        headerName: 'Acciones',
        renderCell: ({ row }: GridValueGetterParams) => {
          return (
            <>
              <Button
                color='primary'
                onClick={() =>  { handleEdit(row.objectId) } }
              >
                <ModeEditOutlineIcon />
              </Button>
              <Button
                color='primary'
                onClick={() =>  { handleDelete(row.objectId) } }
              >
                <DeleteOutlineIcon />
              </Button>
            </>
          )
        },
        width: 200,
      }
  ];

  const handleEdit = (taskObjectId: string) => {
    console.log('Editar tarea: ', taskObjectId);
    dispatch(setActiveTask( taskObjectId ));
    dispatch(setNavigationStatus({ status: 'editTask' }));
    
  }
  const handleDelete = (taskObjectId: string) => {
    console.log('Eliminar tarea id: ', taskObjectId);
    dispatch(setActiveTask( taskObjectId ));
  }

  

  return (
    <>
        <h1>View Task Screen</h1>

        <Grid container className='fadeIn'>
          <Grid item xs={12} sx={{ height:650, width:'100%' }}>
              <DataGrid
                  columns={ columns } 
                  rows={ rows }                    
                  pageSize={10}
                  rowsPerPageOptions={[5, 10, 20, 50]}
              />
          </Grid>
        </Grid>
        
    </>
  )
}
