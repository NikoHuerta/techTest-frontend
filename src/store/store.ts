import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { taskReducer, navigationReducer } from './slices';

export const store = configureStore({
  reducer: {
    task: taskReducer,
    navigation: navigationReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
