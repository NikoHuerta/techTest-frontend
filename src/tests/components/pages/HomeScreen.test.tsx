import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Provider } from 'react-redux';
import { ThemeProvider, CssBaseline } from '@mui/material';
import { SnackbarProvider } from 'notistack';

import { darkTheme } from '../../../themes';
import { store } from '../../../store';
import { HomeScreen } from '../../../components/pages';

describe('Pruebas en HomeScreen.tsx', () => {

    

    const { asFragment, container } = render( 
        <Provider store={ store }>
            <ThemeProvider theme={ darkTheme }>
                <SnackbarProvider>

                    <HomeScreen />

                <CssBaseline />
                </SnackbarProvider>
            </ThemeProvider>
        </Provider>);

    test('Debe de renderizarse correctamente', () => {
        expect(asFragment()).toMatchSnapshot();
    });

    test('Debe de mostrar el titulo', () => {
        expect(container.querySelectorAll('h1')).toBeDefined();
    });

});

