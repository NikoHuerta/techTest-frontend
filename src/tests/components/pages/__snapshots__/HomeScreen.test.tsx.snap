// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en HomeScreen.tsx Debe de renderizarse correctamente 1`] = `
<DocumentFragment>
  <h1>
    Home Screen
  </h1>
  <h6
    class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-8paf5q-MuiTypography-root"
  >
    App encargada de manejar tareas.
  </h6>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-160tic2-MuiTypography-root"
  >
    Para ver las tareas, haga click en el boton "Ver Tareas".
  </h6>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-1oxen2w-MuiTypography-root"
  >
    Para agregar una nueva tarea, haga click en el boton "Nueva Tarea".
  </h6>
  <h6
    class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-1oxen2w-MuiTypography-root"
  >
    En la visualizacion usted podra Editar y Borrar una tarea respectiva.
  </h6>
</DocumentFragment>
`;
